x-pack/plugin/autoscaling/src/main/java/org/elasticsearch/xpack/autoscaling/storage/ReactiveStorageDeciderService.java

```diff
@@ -274,15 +279,26 @@ public static class AllocationState {
             this.roles = roles;
         }
 
+        record DecisionAllocationResult(ShardRouting shard, boolean decision, List<NodeAllocationResult> nodeAllocationResults) {}
+
         public ShardsSize storagePreventsAllocation() {
             RoutingAllocation allocation = new RoutingAllocation(allocationDeciders, state, info, shardSizeInfo, System.nanoTime());
-            List<ShardRouting> unassignedShards = StreamSupport.stream(state.getRoutingNodes().unassigned().spliterator(), false)
+            List<DecisionAllocationResult> unassignedShards = StreamSupport.stream(
+                state.getRoutingNodes().unassigned().spliterator(),
+                false
+            )
                 .filter(shard -> canAllocate(shard, allocation) == false)
-                .filter(shard -> cannotAllocateDueToStorage(shard, allocation))
+                .map(shard -> cannotAllocateDueToStorage(shard, allocation))
+                .filter(r -> r.decision)
                 .toList();
             return new ShardsSize(
-                unassignedShards.stream().mapToLong(this::sizeOf).sum(),
-                unassignedShards.stream().map(ShardRouting::shardId).collect(Collectors.toCollection(TreeSet::new))
+                unassignedShards.stream().map(e -> e.shard).mapToLong(this::sizeOf).sum(),
+                unassignedShards.stream().map(e -> e.shard.shardId()).collect(Collectors.toCollection(TreeSet::new)),
+                unassignedShards.stream()
+                    .filter(e -> e.nodeAllocationResults.size() > 0)
+                    .min(Comparator.comparing(e -> e.shard.shardId()))
+                    .map(e -> e.nodeAllocationResults)
+                    .orElse(null)
```

Comment:
Can't we default here to an empty list when there aren't unassigned shards due to storage?